clean_cap_dump.c

```diff
@@ -0,0 +1,245 @@
+/* CyberReboot edition:
+ *
+ * In new creating sanitization methods for tcpdump, it became necessary
+ * to make modifications to pcap_dump() supplied by pcap/sf-pcap -- and in
+ * so doing, required an awareness of internal structs located in the
+ * pcap-int.h (a component of libpcap). To simplify things, required data
+ * structs have been extrapolated and given new names to avoid conflict,
+ * and while these are completely extricated from libpcap's internal
+ * structure, data types and sizes should be identical (as they've been
+ * shamelessly cut-and-pasted from pcap-int.h).
+ *
+ *
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include <stdlib.h>
+#include <string.h>
+
+#include <netdissect-stdinc.h>
+
+#include <inttypes.h>
+#include "netdissect.h"
+#include <pcap/pcap.h>
+#include <ether.h>
+#include "clean_cap_dump.h"
+#include <extract.h>
+#include "ethertype.h"
+#include "ip.h"
+#include "tcp.h"
+#include "udp.h"
+
+#define MAXPACKET   65549
+
+
+/* Returns the pointer of the IP header or NULL if that couldn't be found. */
+u_char *
+get_iph_ptr(const struct pcap_pkthdr *h, u_char *bp) {
+	u_short length_type;
+	u_char *s;
+	const struct ip *ip;
+	const struct ether_header *ep = (const struct ether_header *)bp;
+
+	if (h->caplen < ETHER_HDRLEN || h->len < ETHER_HDRLEN) {
+		return NULL;
+	}
+	length_type = EXTRACT_16BITS(&ep->ether_length_type);
+	if (length_type == ETHERTYPE_IP) {
+		s = bp + ETHER_HDRLEN;
+	ip = (const struct ip *)s;
+	if (IP_V(ip) == 4)
+		return s;
+	}
+	return NULL;
+}
+
+static uint32_t
+nd_ipv4_to_network_uint(nd_ipv4 a) {
+	char ip[INET_ADDRSTRLEN];
+	uint32_t ip_uint = 0;
+
+	snprintf(ip, INET_ADDRSTRLEN, "%d.%d.%d.%d",
+		 a.bytes[0], a.bytes[1], a.bytes[2], a.bytes[3]);
+	inet_pton(AF_INET, ip, &ip_uint);
+
+	return ip_uint;
+}
+
+static nd_ipv4
+network_uint_to_nd_ipv4(uint32_t a) {
+	nd_ipv4 addr;
+
+	addr.bytes[3] = (a>>24) & 0xff;
+	addr.bytes[2] = (a>>16) & 0xff;
+	addr.bytes[1] = (a>>8) & 0xff;
+	addr.bytes[0] = a & 0xff;
+
+	return addr;
+}
+
+/* Returns 1 if IP address presented in nd_ipv4 falls within reserved IP range;
+ * else returns 0. */
+static int
+is_reserved(nd_ipv4 a) {
+	/* List of all the reserved IPv4 address spaces, per RFC5735,
+	 * ...PROVIDED IN NETWORK BYTE ORDER!! */
+	struct netblock specialblock[] = {
+		{ .netip = 0x00000000, .netmask = 0x000000ff  }, /* 0.0.0.0/8 */
```

Comment:
@gvanem, could you clarify if it is not available because you use a compiler that does not support C99 in principle or because tcpdump runs the compiler with C99 disabled?